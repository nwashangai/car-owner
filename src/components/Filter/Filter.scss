@import 'sass-rem';
@import 'assets/variables';

.filter {
  min-height: rem(180px);
  margin: rem(13px);
  padding: 2%;
  cursor: pointer;

  &__interval {
    width: 100%;
    margin-bottom: rem(10px);
    font-size: rem(30px);
    font-weight: bolder;
    text-align: center;
  }

  &__gender {
    margin: rem(10px) 0;
    font-size: rem(16px);
    text-align: center;
  }

  &__countries,
  &__colors {
    display: flex;
    justify-content: center;
    margin: rem(10px) 0;
    flex-wrap: wrap;
  }
}
