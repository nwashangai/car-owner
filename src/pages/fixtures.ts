export const initialState: any = {
  hasMore: true,
  isLoading: false,
  showRecords: false,
  limit: 20,
  filter: null,
  filterTrack: false,
  filters: [],
  records: []
};

export const filters = [
  {
    id: 1,
    start_year: 2007,
    end_year: 2009,
    gender: 'male',
    countries: ['Brazil', 'Ireland', 'Egypt', 'Peru'],
    colors: ['Green', 'Violet', 'Yellow', 'Blue']
  },
  {
    id: 2,
    start_year: 1990,
    end_year: 2010,
    gender: '',
    countries: ['Russia', 'Portugal', 'Vietnam', 'Croatia', 'Uganda', 'Iran'],
    colors: ['Teal', 'Maroon', 'Red', 'Orange']
  },
  {
    id: 3,
    start_year: 2001,
    end_year: 2009,
    gender: 'female',
    countries: ['Poland', 'Niger', 'Greece'],
    colors: ['Red', 'Aquamarine', 'Orange', 'Mauv']
  },
  {
    id: 4,
    start_year: 1990,
    end_year: 2000,
    gender: '',
    countries: ['United States', 'Portugal'],
    colors: ['Red']
  },
  {
    id: 5,
    start_year: 2000,
    end_year: 2009,
    gender: '',
    countries: [
      'Myanmar',
      'South Africa',
      'Nicaragua',
      'Finland',
      'Cuba',
      'Mexico'
    ],
    colors: []
  },
  {
    id: 6,
    start_year: 2005,
    end_year: 2010,
    gender: '',
    countries: ['South Africa', 'Cuba', 'Niger', 'Egypt', 'Uganda', 'Kenya'],
    colors: [
      'Red',
      'Aquamarine',
      'Orange',
      'Mauv',
      'Green',
      'Violet',
      'Yellow',
      'Blue',
      'Teal',
      'Maroon',
      'Orange'
    ]
  }
];

export const record = {
  id: 65765,
  first_name: 'Cornelius',
  last_name: 'Westbrook',
  email: 'cwestbrook7d@census.gov',
  country: 'Brazil',
  car_model: 'Honda',
  car_model_year: 2009,
  car_color: 'Yellow',
  gender: 'Male',
  job_title: 'Information Systems Manager',
  bio:
    'Maecenas tristique, est et tempus semper, est quam pharetra magna, ac consequat metus sapien ut nunc. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia Curae; Mauris viverra diam vitae quam. Suspendisse potenti. Nullam porttitor lacus at turpis. Donec posuere metus vitae ipsum.',
  createdAt: '2020-06-06T22:56:37.931Z',
  updatedAt: '2020-06-06T22:56:37.931Z'
};
