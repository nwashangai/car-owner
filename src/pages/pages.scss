@import 'sass-rem';
@import 'assets/variables';

.pages {
  position: relative;

  &__filters {
    display: flex;
    flex-direction: column;
  }

  &__no-filter,
  &__no-record {
    margin: rem(30px) auto;
    font-size: rem(20px);
    font-weight: bolder;
    text-align: center;
    color: map-get($grey, 500);
  }

  &__load-more {
    align-items: center;
    text-align: center;
  }

  &__filter-icon,
  &__records-icon {
    display: flex;

    span {
      margin: auto;
      margin-left: rem(10px);
      font-size: rem(20px);
    }
  }

  &__spinner {
    position: absolute;
    top: 50%;
    left: 50%;
    width: fit-content;
    height: fit-content;
    transform: translate(-50%, -50%);

    &__loader {
      width: rem(150px);
      height: rem(150px);
      border: rem(6px) solid map-get($grey, 500);
      border-right-color: transparent;
    }
  }
}
