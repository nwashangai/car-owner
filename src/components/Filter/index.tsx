// react libraries
import * as React from 'react';

// Components
import Card from 'components/Card';
import Color from 'components/Color';
import Tag from 'components/Tag';

// Styles
import './Filter.scss';

// const countries = [
//   'Russia',
//   'Portugal',
//   'Vietnam',
//   'Croatia',
//   'Uganda',
//   'Iran'
// ];

// const colors = [
//   'Red',
//   'Aquamarine',
//   'Orange',
//   'Mauv',
//   'Green',
//   'Violet',
//   'Yellow',
//   'Blue',
//   'Teal',
//   'Maroon',
//   'Orange'
// ];

export default ({ filter, onClick }: any) => (
  <Card classes="card filter" attributes-id={filter.id} onClick={onClick}>
    <div className="filter__interval">
      <span>
        {filter.start_year ? filter.start_year : '*'} -{' '}
        {filter.end_year ? filter.end_year : '*'}
      </span>
    </div>
    <div className="filter__gender">
      {filter.gender
        ? filter.gender.charAt(0).toUpperCase() +
          filter.gender.slice(1).toLowerCase()
        : ''}
    </div>
    <div className="filter__countries">
      {filter.countries.map((country: string, ind: number) => (
        <Tag tag={country} key={ind} />
      ))}
    </div>
    <div className="filter__colors">
      {filter.colors.map((color: string, ind: number) => (
        <Color color={color} key={ind} />
      ))}
    </div>
  </Card>
);
