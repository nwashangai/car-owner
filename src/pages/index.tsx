// react libraries
import * as React from 'react';
import Icon from '@mdi/react';
import { mdiFilterVariant, mdiFileDocument } from '@mdi/js';
import InfiniteScroll from 'react-infinite-scroll-component';

// Components
import Filter from 'components/Filter';
import Record from 'components/Record';
import Spinner from 'components/Spinner';

// utilities
import getRequest from 'utilities/getRequest';

// Initaial state
import { initialState } from './fixtures';

// Styles
import './pages.scss';

export default () => {
  const [state, setState] = React.useState(initialState);

  React.useEffect(() => {
    setState({
      ...state,
      isLoading: true
    });
    getRequest(`${process.env.REACT_APP_FILTER_API}`)
      .then((response: any) => {
        setState({
          ...state,
          isLoading: false,
          filters: [...response]
        });
      })
      .catch((error: any) => {
        console.log(error.message);
        setState({
          ...state,
          isLoading: false
        });
      });
  }, []);

  React.useEffect(() => {
    if (state.filter) {
      getMoreRecords();
    }
  }, [state.filterTrack]);

  const showRecords = (event: any) => {
    const id = event.currentTarget.attributes['attributes-id'].value;
    const filter = state.filters.find((item: any) => item.id == id);
    if (filter) {
      setState({
        ...state,
        filter,
        filterTrack: !state.filterTrack,
        isLoading: true
      });
    }
  };

  const getMoreRecords = () => {
    const body = {
      options: state.filter,
      limit: state.limit,
      offset: state.records.length
    };
    getRequest(`${process.env.REACT_APP_API_BASE_URL}records`, 'post', body)
      .then((response: any) => {
        setState({
          ...state,
          showRecords: true,
          hasMore: response.data.length > 0,
          isLoading: false,
          records: [...state.records, ...response.data]
        });
      })
      .catch((error: any) => {
        setState({
          ...state,
          isLoading: false
        });
      });
  };

  return (
    <>
      {state.isLoading ? (
        <div className="pages__spinner">
          <Spinner classes="pages__spinner__loader" />
        </div>
      ) : (
        <>
          {state.showRecords ? (
            <div className="pages__records">
              <div className="pages__records-icon">
                <Icon
                  path={mdiFileDocument}
                  title="filter"
                  size={1.2}
                  className=""
                />
                <span>Car Owners Record</span>
              </div>
              {state.records.length > 0 ? (
                <InfiniteScroll
                  dataLength={state.records.length}
                  next={() =>
                    setState({ ...state, filterTrack: !state.filterTrack })
                  }
                  hasMore={state.hasMore}
                  loader={
                    <div className="pages__load-more">
                      <Spinner classes="" />
                    </div>
                  }
                  endMessage={
                    <div className="pages__no-record">No More Records</div>
                  }
                >
                  {state.records.map((data: any, ind: number) => (
                    <Record record={data} key={ind} />
                  ))}
                </InfiniteScroll>
              ) : (
                <div className="pages__no-record">No Record available</div>
              )}
            </div>
          ) : (
            <>
              <div className="pages__filters">
                <div className="pages__filter-icon">
                  <Icon
                    path={mdiFilterVariant}
                    title="filter"
                    size={1.2}
                    className=""
                  />
                  <span>Filter</span>
                </div>
                {state.filters.length > 0 ? (
                  <>
                    {state.filters.map((data: any, ind: number) => (
                      <Filter filter={data} onClick={showRecords} key={ind} />
                    ))}
                  </>
                ) : (
                  <div className="pages__no-filter">No filter available</div>
                )}
              </div>
            </>
          )}
        </>
      )}
    </>
  );
};
