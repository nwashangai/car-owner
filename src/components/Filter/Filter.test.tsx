// react libraries
import * as React from 'react';

// third-party libraries
import { mount } from 'enzyme';

// fixtures
import { filters } from '../../pages/fixtures';

// components
import Filter from '.';

describe('The Filter component', () => {
  it('displays Filter', () => {
    const wrapper = mount(<Filter filter={filters[0]} />);

    const child = wrapper.find('.filter__countries');
    expect(child).toMatchSnapshot();
    expect(child.length).toBe(1);
  });
});
