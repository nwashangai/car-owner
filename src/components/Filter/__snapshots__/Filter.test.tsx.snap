// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Filter component displays Filter 1`] = `ReactWrapper {}`;
